### ISSUE TITLE
[$250] [Search v1] RHP reports show incorrect headers

### ISSUE THREAD

**Issue by luacmartins** [Expensify/App#41585]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:**
**Slack conversation:**

## Action Performed:
Coming from [here](https://github.com/Expensify/App/pull/41347#issuecomment-2091956546), the report headers in the RHP are incorrect. They should show the headers for a transaction thread.

## Expected Result:
Reports show incorrect header

## Actual Result:
Reports don't show correct header

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017a5debcfaef3f611</li>
        <li>Upwork Job ID: 1786404294566879232</li>
        <li>Last Price Increase: 2024-07-19</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @fedirjh</details>

**Comment by Niki106** [comment]
Hi,
I'd like to work on this.

**Comment by Niki106** [comment]
Contributor details
nikimijajlovicfl@hotmail.com
https://www.upwork.com/freelancers/~012b26235f47ef40e2

**Comment by Niki106** [comment]
OK, I will do it quickly.

**Comment by rakhaxor** [comment]
Contributor details
Your Expensify account email: funsofts@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01f94219e6101472c1

**Comment by nkdengineer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Reports don't show correct header

### What is the root cause of that problem?
We render the header as `MoneyRequestHeader` if it's a money request by using `isMoneyRequest` function

https://github.com/Expensify/App/blob/2f5173c9c0f041967acfbdf9581ce52327623fa6/src/pages/home/ReportScreen.tsx#L338-L339

In this function, we're checking if the `parentReport` is IOU report and the `parentReportAction` is transaction thread. 

https://github.com/Expensify/App/blob/2f5173c9c0f041967acfbdf9581ce52327623fa6/src/libs/ReportUtils.ts#L1343

But if the IOU report has notification as `hidden`, `OpenApp` API doesn't return this report. So after we logout and login again, the IOU report doesn't exist and then when we open the transaction thread report, the header is rendered as `HeaderView` component that shows the `Join` button.
### What changes do you think we should make in order to solve the problem?
1. In `ReportActionItem`, we only checking the `parentReportAction` is transaction thread report or not to display `MoneyRequestView`.

https://github.com/Expensify/App/blob/2f5173c9c0f041967acfbdf9581ce52327623fa6/src/pages/home/report/ReportActionItem.tsx#L768

So we can do the same for `isIOURequest` and `isExpenseRequest` function here

```
!isEmptyObject(parentReportAction) && ReportActionsUtils.isTransactionThread(parentReportAction);
```
https://github.com/Expensify/App/blob/2f5173c9c0f041967acfbdf9581ce52327623fa6/src/libs/ReportUtils.ts#L1343

2. To display the parent subtitle in header in `MoneyRequestView` or `ReportScreen` we should call `OpenReport` API with reportID is the `parentReportID` of transaction thread report if the parent report doesn't exist.

### What alternative solutions did you explore? (Optional)
NA

**Comment by tsa321** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

RHP search report view displays incorrect headers (different from report screen view in center)

### What is the root cause of that problem?
The search result is generated by API.read(READ_COMMANDS.SEARCH and stored in onyx Snapshot_

When user select search result item, the app will display RHP report view. Based on:

https://github.com/Expensify/App/blob/ab1e924bf0d4f928a9006f5bdd6b9da60346b1df/src/components/Search.tsx#L67-L69

The parameter used to open report is `item.transactionThreadReportID`, then the openReport function will open the thread:

https://github.com/Expensify/App/blob/ab1e924bf0d4f928a9006f5bdd6b9da60346b1df/src/components/Search.tsx#L55

But the actual report thread in the search result is placed in the `reportID` property, so instead of opening thread report the openReport wil open thread of thread of the search result.

This will cause different report header than the real thread of the search result. In reportscreen:

https://github.com/Expensify/App/blob/ab1e924bf0d4f928a9006f5bdd6b9da60346b1df/src/pages/home/ReportScreen.tsx#L361-L370

In this line because of different reportID the `ReportUtils.isMoneyRequestReport(report) || ReportUtils.isInvoiceReport(report)` check will fail and other headerview will be used and will display join button.

### What changes do you think we should make in order to solve the problem?
In:

https://github.com/Expensify/App/blob/ab1e924bf0d4f928a9006f5bdd6b9da60346b1df/src/components/Search.tsx#L68

We could use `item.reportID` instead of `item.transactionThreadReportID`

**Comment by tsa321** [comment]
@fedirjh make invoice request to someone and select a workspace. 
`+` button -> Send invoice -> enter amount -> select a user to send the invoice -> select a workspace.
Then in the search page result click on the invoice request item (the top most item you just created). The join button will be displayed in the RHP report view.

**Comment by nkdengineer** [comment]
> To display the parent subtitle in header in MoneyRequestView or ReportScreen we should call OpenReport API with reportID is the parentReportID of transaction thread report if the parent report doesn't exist.


@fedirjh After fixing the point 1, the header is rendered correct with `MoneyRequestHeader` but the parent subtitle that will bring us to the IOU report still doesn't appear because the IOU report is empty


https://github.com/Expensify/App/blob/942cad969430f17c5477ae519bf6c3508cb761f8/src/components/AvatarWithDisplayName.tsx#L140-L141


So to display this we should call `OpenReport` API with `reportID` is `parentReportID` of the transaction thread report in `ReportScreen` like this

```
useEffect(() => {
    if (!report.parentReportID) {
        return;
    }
    Report.openReport(report.parentReportID);
}, [report.parentReportID])
```

**Comment by tsa321** [comment]
@fedirjh for point 1,

> 1. The issue can be reproduced when navigating to the transaction thread from outside the search page.

Yes of course, because the RHP is just a report screen so it will reflect what is shown by report screen in center. It is same component. The join button is displayed because it is thread of thread of the transaction report (this thread report id is generated by back end) and we don't join the thread yet so the join button is shown. 

You can try this in any other thread, then press three dot menu button and then click `leave`, the join button will be shown if you go back to that thread.

What I am trying to show here is, The Right panel  doesn't use correct report id:

<details> <summary> video: </summary>

https://github.com/Expensify/App/assets/114975543/2b47288c-6937-456b-a175-0087d4a1de25

</details>

As you can see in the video, I think the intended report id should be displayed in right panel is report id displayed in center in this video. 
I have send some messages in  the report, but because the app use `item.transactionThreadReportID` the other report shown instead and there is no reply at all in the report in Right panel. 
Also, as you can see in the url bar the report id shown in the center and right panel is different. 

> 2. The issue is not limited to invoice requests, but is also replicable with other transactions.

For second point could you give me steps to reproduce it for other request?

**Comment by luacmartins** [comment]
>But the actual report thread in the search result is placed in the reportID property, so instead of opening thread report the openReport wil open thread of thread of the search result.

@nkdengineer I'm not sure that I understand this part of your RCA. We want to open the `transactionThreadReportID` so users can view/edit the transaction directly. We do not want to open the iou/expense report, which is what `reportID` is storing. So it seems like the issue comes from the fact that we're opening the `transactionThreadReportID`, but don't have the parent report stored locally, is that right?

**Comment by tsa321** [comment]
@luacmartins it is in my RCA.

Let me confirm with images.

This is transaction thread:
<details><summary>transaction thread</summary>

![Screenshot 2024-05-07 at 07 36 39](https://github.com/Expensify/App/assets/114975543/42e51cd7-68c9-4544-93a5-54dd61f55ab4)

</details>

<br />
This is iou/expense report:
<details><summary>iou/expense report:</summary>

![Screenshot 2024-05-07 at 07 36 20](https://github.com/Expensify/App/assets/114975543/71fb89be-bebb-4e56-8482-3e493f082ede)

</details>

<br />

In Snapshot onyx:
1. For transaction from `+ button` -> `Send invoice` or `Pay someone` ,  the transaction thread id is stored in `reportID` and not in `transactionThreadReportID`.  The `transactionThreadReportID` contains random thread id generated by server that we don't even visit that thread yet.
2. For transaction from `+ button` -> `Submit expense`, the transaction thread is correctly stored in `transactionThreadReportID`.

**Comment by luacmartins** [comment]
>For transaction from + button -> Send invoice or Pay someone , the transaction thread id is stored in reportID and not in transactionThreadReportID. The transactionThreadReportID contains thread id generated by server that we don't even visit that thread yet.

This seems like an issue that needs to be fixed. @fedirjh did you get a chance to go over the proposals yet?

**Comment by luacmartins** [comment]
@fedirjh are we still looking for proposals or are we selecting one of the current proposals?

**Comment by JmillsExpensify** [comment]
Looks like we're still waiting for a successful proposal, though it'd be good to confirm.

**Comment by tsa321** [comment]
@fedirjh could you re-review my proposal.

In Snapshot onyx:
1. For transaction from `+ button` -> `Send invoice` or `Pay someone` ,  the transaction thread id is stored in `reportID` and not in `transactionThreadReportID`.  The `transactionThreadReportID` contains thread id generated by server that we don't even visit that thread yet.

TLDR:

Just change this line into:

https://github.com/Expensify/App/blob/ab1e924bf0d4f928a9006f5bdd6b9da60346b1df/src/components/Search.tsx#L68

into `openReport(item.reportID);` 

try it for these two type of transacations (transaction from `+ button` -> `Send invoice` or `Pay someone`) and see if the join button disappears.

**Comment by nkdengineer** [comment]
> @nkdengineer, can we update the hook to minimize these requests by only fetching the one from the search view?

@fedirjh  I think we can add to call `Report.openReport` for parent report here when we select a row here

https://github.com/Expensify/App/blob/e8ae3c5acedf0e6788dc574c7b6f3043ca37092a/src/components/Search.tsx#L87

**Comment by rezkiy37** [comment]
> ...I think @rezkiy37 Can help us debug it further since he implemented the invoice flow.

I see the work in progress here. If you need any help, please ping me 🙂

**Comment by tsa321** [comment]
> the transaction thread id is stored in reportID and not in transactionThreadReportID
@fedirjh it is from the server, the data is from the server.

**Comment by tsa321** [comment]
@luacmartins yes.

**Comment by nkdengineer** [comment]
> the transaction thread id is stored in reportID and not in transactionThreadReportID

I checked and I think it's incorrect, the `reportID` here is the ID of the expense report for the case send invoice because we only have one transaction so the expense report is displayed as combine report that isn't the transaction thread report.

**Comment by nkdengineer** [comment]
For the header of the send invoice transaction thread, the problem is `isSingleTransactionView` doesn't not cover the case the report is send invoice request.

In the second thought if we remove the check `isIOUReport` in `isIOURequest` and `isExpenseReport` in `isExpenseRequest`, that makes it's wrong because then we will not be able to distinguish between expense and iou request. My idea here is only update `isSingleTransactionView` to check the `parentReportAction` is transaction thread or not like this because as I mentioned in my proposal we also only checking the `parentReportAction` is transaction thread report or not to display `MoneyRequestView`. With this, we also fix the case for send invoice.

```
const isSingleTransactionView = ReportActionsUtils.isTransactionThread(parentReportAction) || ReportUtils.isTrackExpenseReport(report) || ;

```

https://github.com/Expensify/App/blob/8d73db3b6521c33637cd91a14fe942b6b0741acd/src/pages/home/ReportScreen.tsx#L316

@fedirjh What do you think?

**Comment by luacmartins** [comment]
@fedirjh were you able to take a look at the comments above?

**Comment by luacmartins** [comment]
@nkdengineer we can't rely on checking the parent report action because that data is not available in search.

**Comment by nkdengineer** [comment]
@luacmartins When we view a search result, `OpenReport` API will be called and it will contain parent report action data.

https://github.com/Expensify/App/assets/161821005/f4580a0a-aa7c-4649-9e39-5982d8878e21

**Comment by luacmartins** [comment]
That's correct. That being said, I've seen the incorrect header being shown before the call to OpenReport returns a response. How will we handle that case?

**Comment by luacmartins** [comment]
@fedirjh don't we already return that in OpenReport? I just tested opening a transaction thread and the parent IOU action was returned as well
![Uploading Screenshot 2024-06-04 at 4.59.16 PM.png…]()

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by nkdengineer** [comment]
> That's correct. That being said, I've seen the incorrect header being shown before the call to OpenReport returns a response. How will we handle that case?

@luacmartins 

I just checked and saw that:

1. If the transaction thread report doesn't return in `OpenApp` API (notification is hidden), when we open this report skeleton will be displayed before `OpenReport` API is complete so this case works well with `parentReportAction` check

2. If this report is returned in `OpenApp`, I see that `parentReportAction` of this is also return as well so this case also work well

Correct me if I missed st.

**Comment by luacmartins** [comment]
Hmm I see what you mean now. We're already returning the report type in the Search data, I wonder which path would be best:

1. Update `OpenReport` to return the full parent report data
2. Update the header component to use the data stored in the `snapshot_` Onyx key

**Comment by luacmartins** [comment]
I see. Maybe we'd have to modify the OpenReport command to take a `isFromSearch` param so it can include the parent report data as well

**Comment by luacmartins** [comment]
Yea, we might have to. I'm trying to think of a better solution that decouples OpenReport from the Search RHP.

**Comment by luacmartins** [comment]
@fedirjh @nkdengineer @tsa321 could you share steps to reproduce this bug? I can only manage to get the header to flash the `Join` button, but it goes away once the response from OpenReport is saved in Onyx. You can see that in the video below:

https://github.com/Expensify/App/assets/22219519/6902dbc3-5d3d-4d08-bb88-03415e09d43e

[These steps](https://github.com/Expensify/App/issues/41585#issuecomment-2095352139) don't work because `Invoice` reports are not shown in Search

**Comment by luacmartins** [comment]
Maybe, good to know that it has been implemented already :) Can we do something about the flicker? Maybe we gotta wait until data is available in Onyx?

**Comment by luacmartins** [comment]
Still looking for proposals on this one. Maybe we can just show a loading skeleton until the data is available?

**Comment by JmillsExpensify** [comment]
I'm actually fine with showing a skeleton loader. That's a good idea.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Fourth week)

**Comment by luacmartins** [comment]
Nice! Gonna close this one then. Thank you all! Please reopen if needed.
